app/controllers/admin/facilities_controller.rb

```diff
@@ -42,6 +42,24 @@ def destroy
     redirect_to admin_facilities_url, notice: 'Facility was successfully deleted.'
   end
 
+  def upload
+    authorize Facility
+    file = params[:upload_facilities_file]
+    @file_contents = ''
+    @errors = []
+    run_validations_and_read(file) if file.present?
```

Comment:
Split file type and size validation, reading the file and validating it content in to separate functions. You can also use before action to run the file validations.